File: day2/2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"../helpers"
)

func getPolicy(password string) (string, []string, string) {
	pass := strings.Split(password, " ")
	passRange := strings.Split(pass[0], "-")
	char := pass[1][0]
	passString := pass[2]
	return string(char), passRange, string(passString)
}

func xor(a bool, b bool) bool {
	return (a || b) && !(a && b)
}

func p1(data []string) {
	defer helpers.TimeTrack(time.Now())
	count := 0
	for _, password := range data {
		char, passRange, passString := getPolicy(password)
		occurences := strings.Count(passString, char)
		low, _ := strconv.Atoi(passRange[0])
		high, _ := strconv.Atoi(passRange[1])
		if low <= occurences && high >= occurences {
			count++
		}
	}
	fmt.Printf("Part 1: %d ", count)
}

func p2(data []string) {
	defer helpers.TimeTrack(time.Now())
	count := 0
	for _, password := range data {
		char, passRange, passString := getPolicy(password)
		low, _ := strconv.Atoi(passRange[0])
		high, _ := strconv.Atoi(passRange[1])
		a := string(passString[low-1]) == char
		b := string(passString[high-1]) == char
		if xor(a, b) {
			count++
		}
	}
	fmt.Printf("Part 1: %d ", count)
}

func main() {
	fmt.Println("Day 2: Password Philosophy")
	fileName := "/Users/jack/Documents/Github/aoc-2020/day2/2.in"
	data, err := helpers.GetInputStringByLine(fileName)
	if err != nil {
		panic(err)
	}
	p1(data)
	p2(data)
}

File: day2/2.py
from timeit import default_timer as timer

def get_input():
    with open("./day2/2.in", 'r') as f:
        return [line.strip() for line in f.readlines()]

def get_policy(password):
    password =  password.split(' ')
    passRange = password[0].split('-')
    char = password[1][0]
    passString = password[2]
    return char, passRange, passString


def p1():
    valid_passwords = 0
    data = get_input()
    for password in data:
        char, passRange, passString = get_policy(password)
        if passString.count(char) in range(int(passRange[0]), int(passRange[1])+1):
            valid_passwords += 1
    return valid_passwords
    

def p2():
    valid_passwords = 0
    data = get_input()
    for password in data:
        char, passPos, passString = get_policy(password)
        # Only one can be True
        if (passString[int(passPos[0])-1] == char) ^ (passString[int(passPos[1])-1]==char):
            valid_passwords += 1
    return valid_passwords

print("Day 2: Password Philosophy")  

p1start = timer()
p1 = p1()
p1end = timer()
print(f"Part 1: {p1} in {p1end-p1start}s.")

p2start = timer()
p2 = p2()
p2end = timer()
print(f"Part 2: {p2} in {p2end-p2start}s.")
    